File: updater/src/main/java/identifiers/FaceNormal.java
package identifiers;

import org.objectweb.asm.Type;
import utility.ClassWrapper;

public class FaceNormal extends AbstractIdentifier {
    @Override
    public boolean isMatch(ClassWrapper classNode) {
        return classNode.getInterfaces().isEmpty() &&
                classNode.getInstanceFields().size() == 3 &&
                classNode.getInstanceFields().stream().allMatch(fieldNode -> Type.getType(fieldNode.desc).equals(Type.INT_TYPE)) &&
                classNode.getInstanceMethods().isEmpty() &&
                classNode.getConstructors().size() == 1;
        //TODO .and { it.constructors.first().instructions.none { it.isField } }
    }
}

File: find-names/src/main/java/mapper/identifiers/classes/FaceNormal.kt
package mapper.identifiers.classes

import mapper.abstractclasses.IdentityMapper
import mapper.predicateutilities.and
import mapper.predicateutilities.predicateOf
import mapper.predicateutilities.type
import mapper.wrappers.ClassWrapper
import org.objectweb.asm.Type.INT_TYPE

class FaceNormal : IdentityMapper.Class() {

    override val predicate = predicateOf<ClassWrapper> { it.superType == Any::class.type }
            .and { it.interfaces.isEmpty() }
            .and { it.instanceFields.count() == 3 }
            .and { it.instanceFields.all { it.type == INT_TYPE } }
            .and { it.instanceMethods.isEmpty() }
            .and { it.constructors.size == 1 }
            .and { it.constructors.first().instructions.none { it.isField } }
            .and { it.staticMethods.size == 1}
}